tests/sepp/Execution.cpp

```diff
@@ -0,0 +1,146 @@
+      _workload(std::move(workload)) {}
+
+Execution::~Execution() {
+  _state.store(ExecutionState::kStopped);
+  for (auto& thread : _threads) {
+    if (thread->_thread.joinable()) {
+      thread->_thread.join();
+    }
+  }
+}
+
+void Execution::createThreads(Server& server) {
+  _threads = _workload->createThreads(*this, server);
+}
+
+ExecutionState Execution::state(std::memory_order order) const {
+  return _state.load(order);
+}
+
+void Execution::signalFinishedThread() noexcept { _activeThreads.fetch_sub(1); }
+
+Report Execution::run() {
+  _activeThreads.store(_threads.size());
+  _state.store(ExecutionState::kPreparing);
+
+  waitUntilAllThreadsAre(ThreadState::kRunning);
+
+  _state.store(ExecutionState::kInitializing);
+
+  waitUntilAllThreadsAre(ThreadState::kReady);
+
+  _state.store(ExecutionState::kRunning);
+
+  auto start = std::chrono::high_resolution_clock::now();
+
+  std::uint32_t sleeptimePerRound = 100;
+  auto rounds = [&]() -> std::uint32_t {
+    auto stoppingCriterion = _workload->stoppingCriterion();
+    if (std::holds_alternative<StoppingCriterion::Runtime>(stoppingCriterion)) {
+      return std::get<StoppingCriterion::Runtime>(stoppingCriterion).ms /
+             sleeptimePerRound;
+    } else {
+      // for stopping criteria based on number of operations we use an artifical
+      // time limit of one hour
+      return 1000 * 60 * 60 / 100;
+    }
+  }();
+
+  for (std::uint32_t r = 0;
+       r < rounds && _activeThreads.load(std::memory_order_relaxed) > 0; ++r) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(sleeptimePerRound));
+    // TODO - record samples (memory usage, rocksdb stats, ...)
+  }
+
+  _state.store(ExecutionState::kStopped);
+
+  waitUntilAllThreadsAre(ThreadState::kFinished);
+
+  std::chrono::duration<double, std::milli> runtime =
+      std::chrono::high_resolution_clock::now() - start;
+  return buildReport(runtime.count());
+}
+
+Report Execution::buildReport(double runtime) {
+  for (auto& thread : _threads) {
+    if (thread->_thread.joinable()) {
+      thread->_thread.join();
+    }
+  }
+
+  std::vector<ThreadReport> threadReports;
+  threadReports.reserve(_threads.size());
+  for (auto& thread : _threads) {
+    threadReports.push_back(thread->report());
+  }
+  return {.threads = std::move(threadReports),
+          .runtime = runtime,
+          .databaseSize = getFolderSize(_options.databaseDirectory)};
+}
+
+void Execution::waitUntilAllThreadsAre(ThreadState state) {
+  for (auto& thread : _threads) {
+    waitUntilThreadStateIs(*thread, state);
+  }
+}
+
+void Execution::waitUntilThreadStateIs(ExecutionThread const& thread,
+                                       ThreadState expected) {
+  auto state = thread._state.load(std::memory_order_relaxed);
+  while (state != expected) {
+    if (state == ThreadState::kFinished) {
+      throw std::runtime_error("worker thread finished prematurely");
+    }
+    state = thread._state.load(std::memory_order_relaxed);
```

Comment:
Just realized that this will be a busy-wait loop.
That doesn't sound good in general, but I guess we want it anyway for the following call to this function:
```
  _state.store(ExecutionState::kStopped);

  waitUntilAllThreadsAre(ThreadState::kFinished);

  std::chrono::duration<double, std::milli> runtime =
      std::chrono::high_resolution_clock::now() - start;
```
If we would go to sleep inside here, it would influence the timing results.

Reply:
Yes, this as well as the other loops are intentionally busy loops since they should only wait for a rather small time and we want all threads to start at the same time (alltough with a longer runtime small delays during start should not matter much). I can put in a `pause` to reduce the CPU load though...

I would have liked to use a std::barrier instead of the busy loop, but I think some compiler is still lacking support. So a busy loop was the simplest solution (didn't want to implement my own barrier).